File: leetcode/Go/300.go
package main

import "fmt"

func lengthOfLIS(nums []int) int {
	l := len(nums)
	if l == 0 {
		return 0
	}
	var arr = []int{nums[0]}
	for i := 1; i < l; i++ {
		if nums[i] > arr[len(arr)-1] {
			arr = append(arr, nums[i])
		} else {
			index := BinarySearch(arr, nums[i])
			arr[index] = nums[i]
		}
	}

	return len(arr)
}

func BinarySearch(arr []int, n int) int {
	l := len(arr)
	beg := 0
	end := l - 1
	for beg < end {
		mid := (beg + end) / 2
		if n > arr[mid] {
			beg = mid + 1
		} else if n < arr[mid] {
			end = mid
		} else if n == arr[mid] {
			return mid
		}
	}
	return end
}

func main() {
	var arr = []int{1, 3, 6, 7, 9, 10, 13, 16, 18, 21, 24, 26}
	fmt.Println(BinarySearch(arr, 19))
	fmt.Println(lengthOfLIS([]int{10, 9, 2, 5, 3, 7, 101, 18}))
}

File: leetcode/C++/300.cpp
#include <vector>
using namespace std;

class Solution
{
  public:
    int lengthOfLIS(vector<int> &nums)
    {
        if (nums.size() == 0)
            return 0;
        vector<int> dp;
        dp.push_back(nums[0]);
        for (int i = 1; i < nums.size(); ++i)
        {
            if (nums[i] > dp.back())
            {
                dp.push_back(nums[i]);
            }
            else
            {
                int index = binarySearch(dp, nums[i]);
                dp[index] = nums[i];
            }
        }
        return dp.size();
    }
    int binarySearch(vector<int> &nums, int n)
    {
        int beg = 0;
        int end = nums.size() - 1;
        while (beg < end)
        {
            int mid = (beg + end) / 2;
            if (n < nums[mid])
            {
                end = mid;
            }
            else if (n > nums[mid])
            {
                beg = mid + 1;
            }
            else if (n == nums[mid])
            {
                return mid;
            }
        }
        return end;
    }
};

int main()
{
    Solution s;
    vector<int> nums{1, 3, 10, 7, 9, 12, 15, 18};
    //auto r = s.binarySearch(nums, 10);
    auto r = s.lengthOfLIS(nums);
    return 0;
}